src/harness/reference_models/fss_purge/fss_purge.py

```diff
@@ -0,0 +1,332 @@
+  fss_height = fss_record['deploymentParam'][0]['installationParam']['height']
+  fss_height_type = fss_record['deploymentParam'][0]['installationParam']['heightType']
+  fss_azimuth = fss_record['deploymentParam'][0]['installationParam']['antennaAzimuth']
+  fss_pointing_elevation = fss_record['deploymentParam'][0]['installationParam']['elevation']
+  fss_max_gain_dbi = fss_record['deploymentParam'][0]['installationParam']['antennaGain']
+  fss_weight_1 = fss_record['deploymentParam'][0]['weight1']
+  fss_weight_2 = fss_record['deploymentParam'][0]['weight2']
+  # Convert the height to AGL if it was AMSL type.
+  if fss_height_type == 'AMSL':
+     fss_height_agl = convertAmslToAgl(fss_latitude, fss_longitude, fss_height)
+  else:
+     fss_height_agl = fss_height
+
+  # Populate FssProtectionPoint tuple.
+  fss_entity = FssProtectionPoint(latitude=fss_latitude,
+                                  longitude=fss_longitude,
+                                  height_agl=fss_height_agl,
+                                  max_gain_dbi=fss_max_gain_dbi,
+                                  pointing_azimuth=fss_azimuth,
+                                  pointing_elevation=fss_pointing_elevation,
+                                  weight_1=fss_weight_1,
+                                  weight_2=fss_weight_2)
+  return fss_entity
+
+def getNeighbouringCbsdsToFss(cbsd_list, fss_record):
+  """ Get the list of all CBSDs that lie within 40 KMs of FSS.
+
+  Args:
+    cbsd_list: List of CBSDs.
+    fss_record: A FSS record dictionary.
+  Returns:
+    List of CBSDs lying within 40 KMs of FSS.
+  """
+
+  neighbouring_cbsds = []
+  for cbsd in cbsd_list:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['deploymentParam'][0]['installationParam']['latitude'],
+        fss_record['deploymentParam'][0]['installationParam']['longitude'],
+        cbsd['registrationRequest']['installationParam']['latitude'],
+        cbsd['registrationRequest']['installationParam']['longitude'])
+
+    if distance_km <= 40:
+        neighbouring_cbsds.append(cbsd)
+  return neighbouring_cbsds
+
+def db2Linear(x):
+  """ Convert dB to linear scale. """
+  return 10**(x/10.)
+
+def performPurge(cbsds, fss_entity, fss_oobe_margin):
+  """ Implementation of algorithm described in R2-SGN-29.
+
+  The list of grants to be purged based out of band emission are identifies and the grants
+  are removed from the CBSDs. If a CBSDs has no grants then the CBSD is removed from the list.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighbourhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+    fss_oobe_margin: OOBE Margin value to be considered for FSS purge list calculation.
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)

```

Comment:
Are you sure it's safe to modify the input?

Reply:
The   'oobe_interference' parameter that is added to the input is removed at the   end of this function. The addition and removal of this parameter is done   within the performPurge method. This parameter makes the algorithm easier to   implement without the need for an additional list and mapping to the input.   We think it should be ok because the parameter is finally removed and the   input values remain the same before the fss purge model is invoked. Please   let us know if you would like logic to be changed.